testing/txsim/run_test.go

```diff
@@ -0,0 +1,156 @@
+		expMessages map[string]int64
+	}{
+		{
+			name:      "send sequence",
+			sequences: []txsim.Sequence{txsim.NewSendSequence(2, 1000)},
+			// we expect at least 5 bank send messages within 30 seconds
+			expMessages: map[string]int64{sdk.MsgTypeURL(&bank.MsgSend{}): 5},
+		},
+		{
+			name:      "stake sequence",
+			sequences: []txsim.Sequence{txsim.NewStakeSequence(1000)},
+			expMessages: map[string]int64{
+				sdk.MsgTypeURL(&staking.MsgDelegate{}):                     1,
+				sdk.MsgTypeURL(&distribution.MsgWithdrawDelegatorReward{}): 5,
+				// NOTE: this sequence also makes redelegations but because the
+				// testnet has only one validator, this never happens
+			},
+		},
+		{
+			name: "blob sequence",
+			sequences: []txsim.Sequence{
+				txsim.NewBlobSequence(
+					txsim.NewRange(100, 1000),
+					txsim.NewRange(1, 3)),
+			},
+			expMessages: map[string]int64{sdk.MsgTypeURL(&blob.MsgPayForBlobs{}): 10},
+		},
+		{
+			name: "multi blob sequence",
+			sequences: txsim.NewBlobSequence(
+				txsim.NewRange(100, 1000),
+				txsim.NewRange(1, 3),
+			).Clone(4),
+			expMessages: map[string]int64{sdk.MsgTypeURL(&blob.MsgPayForBlobs{}): 20},
+		},
+		{
+			name: "multi mixed sequence",
+			sequences: append(append(
+				txsim.NewSendSequence(2, 1000).Clone(3),
+				txsim.NewStakeSequence(1000).Clone(3)...),
+				txsim.NewBlobSequence(txsim.NewRange(100, 400), txsim.NewRange(1, 4)).Clone(2)...),
+			expMessages: map[string]int64{
+				sdk.MsgTypeURL(&bank.MsgSend{}):                            15,
+				sdk.MsgTypeURL(&staking.MsgDelegate{}):                     2,
+				sdk.MsgTypeURL(&distribution.MsgWithdrawDelegatorReward{}): 10,
+				sdk.MsgTypeURL(&blob.MsgPayForBlobs{}):                     10,
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+
+			keyring, rpcAddr, grpcAddr := Setup(t)
+
+			err := txsim.Run(
+				ctx,
+				[]string{rpcAddr},
+				[]string{grpcAddr},
+				keyring,
+				9001,
+				time.Second,
+				tc.sequences...,
+			)
+			// Expect all sequences to run for at least 30 seconds without error
+			require.True(t, errors.Is(err, context.DeadlineExceeded), err.Error())
+
+			blocks, err := testnode.ReadBlockchain(context.Background(), rpcAddr)
+			require.NoError(t, err)
+			for _, block := range blocks {
+				msgs, err := testnode.DecodeBlockData(block.Data)
+				require.NoError(t, err, block.Height)
+				for _, msg := range msgs {
+					if _, ok := tc.expMessages[sdk.MsgTypeURL(msg)]; ok {
+						tc.expMessages[sdk.MsgTypeURL(msg)]--
+					}
+				}
+			}
+			for msg, count := range tc.expMessages {
+				if count > 0 {
+					t.Errorf("missing %d messages of type %s (blocks: %d)", count, msg, len(blocks))
+				}
+			}
+		})
+	}
+}
+
+func Setup(t *testing.T) (keyring.Keyring, string, string) {
+	t.Helper()
+	genesis, keyring, err := testnode.DefaultGenesisState()
```

Comment:
[not blocking]
the funding of the master account is very implicit atm, imo we should document that here or fund an explicit account ahead of time

Reply:
Good Idea. I will write something up